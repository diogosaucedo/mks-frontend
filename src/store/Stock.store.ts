import { PayloadAction, createSlice } from '@reduxjs/toolkit';

import { ICardProduct } from '@/interfaces';

const stock = createSlice({
  name: 'stock',
  initialState: {
    counter: 0,
    total: 0,
    visible: false,
    products: [] as never | ICardProduct[],
  },
  reducers: {
    toggleVisible(state) {
      state.visible = !state.visible;
    },
    addProduct(state, action: PayloadAction<ICardProduct>) {
      if (
        state.products.find(
          (item: ICardProduct) => item.id == action.payload.id,
        )
      ) {
      } else {
        state.products.push(action.payload);
      }
    },
    removeProduct(state, action: PayloadAction<number>) {
      state.products = state.products.filter(
        (product) => product.id !== action.payload,
      );
    },
    counter(state) {
      state.counter = state.products.length;
    },
    getTotal(state) {
      state.total = state.products.reduce(
        (acc, item) => acc + item.amount * item.price,
        0,
      );
    },
    incrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload) {
          return { ...product, amount: product.amount + 1 };
        } else {
          return product;
        }
      });
    },
    decrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload && product.amount > 1) {
          return { ...product, amount: product.amount - 1 };
        } else {
          return product;
        }
      });
    },
  },
});

export const {
  toggleVisible,
  addProduct,
  counter,
  removeProduct,
  getTotal,
  incrementAmountProduct,
  decrementAmountProduct,
} = stock.actions;
export default stock.reducer;
