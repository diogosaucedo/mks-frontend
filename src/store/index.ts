import { configureStore } from '@reduxjs/toolkit'
import stockStore from './Stock.store'

export const store = configureStore({
  reducer: {
    stock: stockStore
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
