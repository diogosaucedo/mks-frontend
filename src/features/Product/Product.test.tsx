import { store } from '@/store/';
import { addProduct, counter, getTotal } from '@/store/Stock.store';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider, useDispatch } from 'react-redux';

import { Product } from '.';

jest.mock('react-redux', () => {
  const originalModule = jest.requireActual('react-redux');

  return {
    ...originalModule,
    useDispatch: jest.fn().mockReturnValue(jest.fn()),
  };
});

describe('Product feature', () => {
  it('should render properties correctly', () => {
    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      brand: 'HyperX',
      description:
        'O HyperX Cloud Stinger™ é o headset ideal para jogadores que procuram leveza e conforto, qualidade de som superior e maior praticidade.',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      createdAt: '2023-01-23T18:17:04.771Z',
      updatedAt: '2023-01-23T18:17:04.771Z',
    };
    const truncatedPrice = Math.trunc(product.price);
    render(
      <Provider store={store}>
        <Product key={product.id} data={product} />
      </Provider>,
    );

    expect(screen.getByText(product.name)).toBeInTheDocument();
    expect(screen.getByText(product.description)).toBeInTheDocument();
    expect(screen.getByText(`R$${truncatedPrice}`)).toBeInTheDocument();
    expect(screen.getByAltText(product.name)).toBeInTheDocument();
    expect(screen.getByAltText(product.name)).toHaveAttribute(
      'src',
      product.photo,
    );
    expect(screen.getByText('COMPRAR')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      brand: 'HyperX',
      description:
        'O HyperX Cloud Stinger™ é o headset ideal para jogadores que procuram leveza e conforto, qualidade de som superior e maior praticidade.',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      createdAt: '2023-01-23T18:17:04.771Z',
      updatedAt: '2023-01-23T18:17:04.771Z',
    };

    const { asFragment } = render(
      <Provider store={store}>
        <Product key={product.id} data={product} />
      </Provider>,
    );

    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('should dispatch addProduct, counter and getTotal actions on button click', () => {
    const dispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(dispatch);
    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      brand: 'HyperX',
      description:
        'O HyperX Cloud Stinger™ é o headset ideal para jogadores que procuram leveza e conforto, qualidade de som superior e maior praticidade.',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      createdAt: '2023-01-23T18:17:04.771Z',
      updatedAt: '2023-01-23T18:17:04.771Z',
    };

    render(
      <Provider store={store}>
        <Product key={product.id} data={product} />
      </Provider>,
    );
    const button = screen.getByRole('button');
    button.click();

    expect(dispatch).toHaveBeenCalledTimes(3);
    expect(dispatch).toHaveBeenCalledWith(
      addProduct({ ...product, amount: 1 }),
    );
    expect(dispatch).toHaveBeenCalledWith(counter());
    expect(dispatch).toHaveBeenCalledWith(getTotal());
  });
});
