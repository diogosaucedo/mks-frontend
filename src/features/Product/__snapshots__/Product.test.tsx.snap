// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component should match snapshot 1`] = `
<div
  class="sc-kDvujY kPajjf sc-ipEyDJ dvLNvM"
>
  <div
    class="sc-gKPRtg VeQuV"
  >
    <img
      alt="Headset Cloud Stinger"
      class="sc-ftTHYK jBOThg"
      height="138px"
      src="https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp"
      width="fit-content"
    />
    <div
      class="sc-gKPRtg ivrkOC"
      width="100%"
    >
      <h2
        class="sc-bcXHqe jHaUzF"
        color="#2C2C2C"
        font-size="16px"
        font-weight="400"
        height="38px"
        type="h2"
        width="120px"
      >
        Headset Cloud Stinger
      </h2>
      <p
        class="sc-gswNZR fiTlfy sc-dkrFOg dMUiud"
        color="#FFFFFF"
        font-size="15px"
        font-weight="700"
      >
        R$600
      </p>
    </div>
    <p
      class="sc-gswNZR dEUSgy"
      font-size="10px"
      font-weight="300"
      height="25px"
      overflow="hidden"
      width="190px"
    >
      O HyperX Cloud Stinger™ é o headset ideal para jogadores que procuram leveza e conforto, qualidade de som superior e maior praticidade.
    </p>
  </div>
  <button
    class="sc-hLBbgP dRxJKH"
    height="32px"
    width="100%"
  >
    <div
      class="sc-jSUZER hIaACa"
    >
      <img
        alt="Shopping bag icon"
        class="sc-ftTHYK fHavDg"
        height="fit-content"
        src="shopping-bag.svg"
        width="13px"
      />
      <p
        class="sc-gswNZR gtgrOE"
        color="#FFFFFF"
        font-size="14px"
        font-weight="600"
      >
        COMPRAR
      </p>
    </div>
  </button>
</div>
`;

exports[`Product feature should match snapshot 1`] = `
<div
  class="sc-kDvujY kPajjf sc-ipEyDJ dvLNvM"
>
  <div
    class="sc-gKPRtg VeQuV"
  >
    <img
      alt="Headset Cloud Stinger"
      class="sc-ftTHYK jBOThg"
      height="138px"
      src="https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp"
      width="fit-content"
    />
    <div
      class="sc-gKPRtg ivrkOC"
      width="100%"
    >
      <h2
        class="sc-bcXHqe jHaUzF"
        color="#2C2C2C"
        font-size="16px"
        font-weight="400"
        height="38px"
        type="h2"
        width="120px"
      >
        Headset Cloud Stinger
      </h2>
      <p
        class="sc-gswNZR fiTlfy sc-dkrFOg dMUiud"
        color="#FFFFFF"
        font-size="15px"
        font-weight="700"
      >
        R$600
      </p>
    </div>
    <p
      class="sc-gswNZR dEUSgy"
      font-size="10px"
      font-weight="300"
      height="25px"
      overflow="hidden"
      width="190px"
    >
      O HyperX Cloud Stinger™ é o headset ideal para jogadores que procuram leveza e conforto, qualidade de som superior e maior praticidade.
    </p>
  </div>
  <button
    class="sc-hLBbgP dRxJKH"
    height="32px"
    width="100%"
  >
    <div
      class="sc-jSUZER hIaACa"
    >
      <img
        alt="Shopping bag icon"
        class="sc-ftTHYK fHavDg"
        height="fit-content"
        src="shopping-bag.svg"
        width="13px"
      />
      <p
        class="sc-gswNZR gtgrOE"
        color="#FFFFFF"
        font-size="14px"
        font-weight="600"
      >
        COMPRAR
      </p>
    </div>
  </button>
</div>
`;
